src/logplex_tlssyslog_drain.erl

```diff
@@ -0,0 +1,742 @@
+               'requests_sent',
+               non_neg_integer(), #state{}) -> any().
+msg_stat(Key, N,
+         #state{drain_id=DrainId, channel_id=ChannelId}) ->
+    logplex_stats:incr(#drain_stat{drain_id=DrainId,
+                                   channel_id=ChannelId,
+                                   key=Key}, N).
+
+-spec duration(#state{}) -> iolist().
+duration(#state{connect_time=undefined}) ->
+    "undefined";
+duration(#state{connect_time=T0}) ->
+    US = timer:now_diff(os:timestamp(), T0),
+    io_lib:format("~f", [US / 1000000]).
+
+%% -spec buffer_status(#state{}) -> 'empty' | 'has_messages_to_send'.
+%% %% @private
+%% buffer_status(State = #state{buf = Buf}) ->
+%%     case logplex_msg_buffer:len(Buf) of
+%%         0 -> empty;
+%%         _ -> has_messages_to_send
+%%     end.
+
+-spec buffer(any(), #state{}) -> #state{}.
+%% @private
+buffer(Msg, State = #state{buf = Buf}) ->
+    {Result, NewBuf} = logplex_msg_buffer:push_ext(Msg, Buf),
+    msg_stat(drain_buffered, 1, State),
+    case Result of
+        displace ->
+            msg_stat(drain_dropped, 1, State),
+            %% If socket is online, note drain dropped.
+            case State#state.sock of
+                undefined -> ok;
+                _ -> logplex_realtime:incr('drain.dropped')
+            end;
+        insert -> ok
+    end,
+    State#state{buf=NewBuf}.
+
+%% @private
+%% @doc Send buffered messages.
+-spec send(#state{}) -> {next_state, 'sending' | 'ready_to_send', #state{}}.
+send(State = #state{buf = Buf, sock = Sock,
+                    drain_tok = DrainTok}) ->
+    case logplex_msg_buffer:empty(Buf) of
+        empty ->
+            {next_state, ready_to_send, State};
+        not_empty ->
+            PktSize = target_send_size(),
+            {Data, _N, NewBuf} =
+                buffer_to_pkts(Buf, PktSize, DrainTok),
+            try
+                %% ssl:send({sslsocket, _, Pid}, Data)
+                %% ssl_connection(Pid, Data)
+                %% gen_fsm:sync_send_all_state_event(Pid, {application_data, iolist_to_binary(Data)})
+                %% Ref = erlang:monitor(Pid)
+                %% erlang:send(Pid, {'$gen_sync_all_state_event', {self(), Ref}, {application_data, iolist_to_binary(Data)}})
+                %% {Ref, Reply} ->
+                %%     erlang:demonitor(Ref, [flush]),
+                %%     {ok, Reply};
+                %% {'DOWN', Ref, _, _, noconnection} ->
+                %%     Node = get_node(Process),
+                %%     exit({nodedown, Node});
+                %% {'DOWN', Ref, _, _, Reason} ->
+                %%     exit(Reason)
+                {sslsocket, _, Pid} = Sock,
+                MRef = erlang:monitor(process, Pid),
+                erlang:send(Pid,
+                            {'$gen_sync_all_state_event',
+                             {self(), MRef},
+                             {application_data, iolist_to_binary(Data)}}),
```

Comment:
Unfortunately yeah - the process design for tcp and now TLS drains requires that no blocking operations happen in the process handling the log channel buffer (which is combined with the drain itself for these two drain types - unlike the http drain). The SSL library doesn't expose an async send + get-message-on-result API, so we dig into the internals to get an async message interface.

The alternative is to split the sender and message buffering. That may be an easier to maintain approach.


> On Jul 23, 2015, at 15:52, Alex Arnell <notifications@github.com> wrote:
> 
> In src/logplex_tlssyslog_drain.erl:
> 
> > +                %% Ref = erlang:monitor(Pid)
> > +                %% erlang:send(Pid, {'$gen_sync_all_state_event', {self(), Ref}, {application_data, iolist_to_binary(Data)}})
> > +                %% {Ref, Reply} ->
> > +                %%     erlang:demonitor(Ref, [flush]),
> > +                %%     {ok, Reply};
> > +                %% {'DOWN', Ref, _, _, noconnection} ->
> > +                %%     Node = get_node(Process),
> > +                %%     exit({nodedown, Node});
> > +                %% {'DOWN', Ref, _, _, Reason} ->
> > +                %%     exit(Reason)
> > +                {sslsocket, _, Pid} = Sock,
> > +                MRef = erlang:monitor(process, Pid),
> > +                erlang:send(Pid,
> > +                            {'$gen_sync_all_state_event',
> > +                             {self(), MRef},
> > +                             {application_data, iolist_to_binary(Data)}}),
> Is there a reason why this monitoring and intimate knowledge of the bowels of the ssl connection logic is required here?
> 
> —
> Reply to this email directly or view it on GitHub.
>